kexoqs.c

```diff
@@ -238,152 +194,50 @@ oqs_ssh2_reply_msg(const OQS_ALG *oqs_alg) {
-}
-
-/*
- * @brief Do nothing function to satisfy OQS_RAND struct
- */
-static void
-oqs_rand_openssh_buf_free(OQS_RAND *oqs_rnd) {
-
-	(void)oqs_rnd;
-}
-
-/*
- * @brief Initialises OQS_RAND struct
- * Use OpenSSH provided PRNG.
- * This is necessary when the sshd daemon runs in sandbox mode.
- */
-static void
-oqs_rand_openssh_buf_init() {
-
-	oqs_rnd.method_name = "arc4random_buf";
-	oqs_rnd.estimated_classical_security = 128;
-	oqs_rnd.estimated_quantum_security = 64; // Grover search
-	/* Use PRNG provided by OpenSSH instad of liboqs's PRNG */
-	oqs_rnd.rand_8 = &oqs_rand_openssh_buf_8;
-	oqs_rnd.rand_32 = &oqs_rand_openssh_buf_32;
-	oqs_rnd.rand_64 = &oqs_rand_openssh_buf_64;
-	oqs_rnd.rand_n = &oqs_rand_openssh_buf_n;
-	oqs_rnd.free = &oqs_rand_openssh_buf_free;
-}
-
 /*
  * @brief Generates the client side part of the liboqs kex
  */
 int
 oqs_client_gen(OQS_KEX_CTX *oqs_kex_ctx) {
 
-	OQS_KEX *oqs_kex = NULL;
+	OQS_KEM *oqs_kem = NULL;
 	int r = 0;
 
-	if (OQS_NEED_SEED == oqs_kex_ctx->oqs_need_seed) {
-		if ((oqs_kex_ctx->oqs_seed = calloc(sizeof(*oqs_kex_ctx->oqs_seed),
-			oqs_kex_ctx->oqs_need_seed_len)) == NULL) {
-			r = SSH_ERR_ALLOC_FAIL;
-			goto out;
-		}
+	if ((oqs_kem = OQS_KEM_new(oqs_kex_ctx->oqs_method)) == NULL) {
+		r = SSH_ERR_INTERNAL_ERROR;
+		goto out;
+	}
 
-		oqs_rand_openssh_buf_n(&oqs_rnd, oqs_kex_ctx->oqs_seed,
-			oqs_kex_ctx->oqs_need_seed_len);
-		oqs_kex_ctx->oqs_seed_len = oqs_kex_ctx->oqs_need_seed_len;
+	oqs_kex_ctx->oqs_local_priv = NULL;
+	oqs_kex_ctx->oqs_local_msg = NULL;
 
-		if (oqs_kex_ctx->oqs_seed == NULL ||
-			oqs_kex_ctx->oqs_seed_len != oqs_kex_ctx->oqs_need_seed_len) {
-			r = SSH_ERR_INTERNAL_ERROR;
-			goto out;
-		}
-	}
-	else {
-		oqs_kex_ctx->oqs_seed = NULL;
-		oqs_kex_ctx->oqs_seed_len = 0;
+	oqs_kex_ctx->oqs_local_priv_len = oqs_kem->length_secret_key;
+	if ((oqs_kex_ctx->oqs_local_priv = malloc(oqs_kem->length_secret_key)) == NULL) {
+		r = SSH_ERR_ALLOC_FAIL;
+		goto out;
 	}
 
-	if ((oqs_kex = OQS_KEX_new(&oqs_rnd, oqs_kex_ctx->oqs_method,
-		oqs_kex_ctx->oqs_seed, oqs_kex_ctx->oqs_seed_len,
-		oqs_kex_ctx->oqs_param)) == NULL) {
-		r = SSH_ERR_INTERNAL_ERROR;
+	oqs_kex_ctx->oqs_local_msg_len = oqs_kem->length_public_key;
+	if ((oqs_kex_ctx->oqs_local_msg = malloc(oqs_kem->length_public_key)) == NULL) {
+		r = SSH_ERR_ALLOC_FAIL;
 		goto out;
 	}
 
 	/* Generate client side part of kex */
-	if (OQS_KEX_alice_0(oqs_kex, &(oqs_kex_ctx->oqs_local_priv),
-		&(oqs_kex_ctx->oqs_local_msg),
-		&(oqs_kex_ctx->oqs_local_msg_len)) != OQS_SUCCESS) {
+	if (OQS_KEM_keypair(oqs_kem, oqs_kex_ctx->oqs_local_msg,
+		oqs_kex_ctx->oqs_local_priv) != OQS_SUCCESS) {
 		r = SSH_ERR_INTERNAL_ERROR;
 		goto out;
 	}
 
-	oqs_kex_ctx->oqs_kex = oqs_kex;
-	oqs_kex = NULL;
+	oqs_kex_ctx->oqs_kem = oqs_kem;
+	oqs_kem = NULL;
 
 out:
-	if (oqs_kex != NULL) {
-		OQS_KEX_free(oqs_kex);
+	if (oqs_kem != NULL) {
+		OQS_KEM_free(oqs_kem);
+		free(oqs_kex_ctx->oqs_local_priv);
+		free(oqs_kex_ctx->oqs_local_msg);
```

Comment:
You can potentially free a NULL pointer here and in the line above.

Reply:
Calling free on NULL is not a problem, it will just do nothing.